core/src/main/java/net/opentsdb/aura/metrics/core/downsample/Aggregator.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * This file is part of OpenTSDB.
+ * Copyright (C) 2021  Yahoo.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.opentsdb.aura.metrics.core.downsample;
+
+import java.util.Arrays;
+import java.util.Iterator;
+
+public abstract class Aggregator {
+
+  protected final double identity;
+  protected double value;
+  protected byte id;
+  protected String name;
+  protected final double[] values;
+
+  private final Aggregator previous;
+  private Aggregator next;
+
+  private AggregatorIterator iterator;
+
+  public static AggregatorBuilder newBuilder() {
+    return new AggregatorBuilder();
+  }
+
+  public Aggregator(final double identity, final byte id, final String name, final int numPoints) {
+    this(identity, id, name, numPoints, null);
+  }
+
+  public Aggregator(
+      final double identity,
+      final byte id,
+      final String name,
+      final int numPoints,
+      final Aggregator previous) {
+    this.identity = identity;
+    this.previous = previous;
+    this.values = new double[numPoints];
+    this.id = id;
+    this.name = name;
+    if (previous != null) {
+      if ((id & previous.id) != 0) {
+        throw new IllegalArgumentException("Duplicate aggregator found for: " + name);
+      }
+      this.id |= previous.id;
```

Comment:
Need some docs around why there is a previous here and why we're doing some bitwise manipulations on the ID. Do we need a linked list of aggs? We could, now, actually re-use the Aggs in the TSD code and keep it simpler.

Reply:
The design enables us to dynamically compose the aggregators. Otherwise, you will have to create classes for all possible combinations of the aggregation. For example `sum, count`, `sum, count, min, max`, `avg` etc.